spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala

```diff
@@ -448,10 +449,93 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
     this
   }
 
+  @deprecated("Please use clone instead", "0.5.0")
   def cloneModule(): AbstractModule[A, B, T] = {
     SerializationUtils.clone(this)
   }
 
+  def clone(deepCopy : Boolean = true): AbstractModule[A, B, T] = {
+    val moduleData = ModuleData[T](this.
+      asInstanceOf[AbstractModule[Activity, Activity, T]], Seq[String](), Seq[String]())
+    val storages = new mutable.HashMap[Int, Any]()
+    val context = SerializeContext(moduleData, storages, ProtoStorageType, false)
+    val serializedModule = ModuleSerializer.serialize[T](context)
+      .bigDLModule
+    ModulePersister.setTensorStorage(serializedModule, storages)
+
+    storages.clear()
+
+    val deserializeContext = DeserializeContext(serializedModule.build,
+      storages, ProtoStorageType, false)
+    ModuleLoader.initTensorStorage[T](deserializeContext)
+    val copy = ModuleSerializer.load[T](deserializeContext).module
+      .asInstanceOf[AbstractModule[A, B, T]]
+    setWeightAndBias(copy, deepCopy)
+    copy
+  }
+
+
+  private def resetStorage(storages: mutable.HashMap[Int, Any]):
+    mutable.HashMap[Int, Any] = {
+    val storages = new mutable.HashMap[Int, Any]()
+    storages
+  }
+
+  private def setWeightAndBias(copy : AbstractModule[A, B, T], deepCopy : Boolean): Unit = {
+    val parameterTable = this.getParametersTable
+    val copiedModuleParamTable = copy.getParametersTable
+    if (parameterTable != null) {
+      require(copiedModuleParamTable != null, "cloned module should have params")
+      parameterTable.foreach {
+        case (name: String, params: Table) =>
+          require(copiedModuleParamTable.get(name) != None, s"cloned module should have for $name")
+          setLayerWeightAndBias(params,
+            copiedModuleParamTable.get(name).get.asInstanceOf[Table], deepCopy)
+      }
+    }
+  }
+
+  private def setLayerWeightAndBias(params : Table,
+                                    copyParams : Table, deepCopy : Boolean): Unit = {
+
+    copyParam(params, copyParams, deepCopy, "weight")
```

Comment:
Maybe there are parameters other than "weight" and "bias" like the "runningMean" and "runningVar" in BatchNormalization